File: control-plane/worker/tests/support/test_database.py
"""The one place a worker test learns which database it may write to.

The Python half of ``control-plane/api/test/support/test-database.ts``, and it
exists for the same reason. These tests do not drop schemas — they insert trips,
bindings and jobs — so pointed at the dev stack's database they do not destroy
it, they contaminate it. That has already happened once: an errored setUp left
fifteen ``prov-test-*`` trips behind in the shared database, which then had to
be found and cleaned by hand.

The rule matches the TypeScript guard exactly: the database NAME must say it is
for tests. Two doors onto the same hazard should not have two different rules,
or the next person will learn one of them and trust it at the other.
"""

from __future__ import annotations

import os
from urllib.parse import urlparse, unquote

# Kept identical to TEST_NAME_PATTERN in the TypeScript guard.
_TEST_NAME = "test"

_GUIDANCE = (
    'CONTROL_PLANE_TEST_DATABASE_URL names the database "{name}", which does '
    "not look like a scratch database.\n\n"
    "These tests write trips, bindings and jobs into whatever database they "
    "are given, so pointing them at a real one contaminates it. The dev "
    "stack's own database is exactly the mistake this catches.\n\n"
    "Use the scratch database instead:\n"
    '  CONTROL_PLANE_TEST_DATABASE_URL="postgres://postgres:test@127.0.0.1:5434/cptest"\n\n'
    "Or leave it unset to skip the database-backed tests."
)


class UnsafeTestDatabaseError(RuntimeError):
    """Raised when the configured database is not safe to write test rows into."""


def database_name_of(connection_string: str) -> str | None:
    """The database name in a Postgres URL, or None when there is not one."""
    try:
        parsed = urlparse(connection_string)
    except ValueError:
        return None
    name = (parsed.path or "").lstrip("/")
    return unquote(name) or None


def is_test_database_url(connection_string: str) -> bool:
    """Whether this URL names a database these tests may write to."""
    name = database_name_of(connection_string)
    return name is not None and _TEST_NAME in name.lower()


def test_database_url(raw: str | None = None) -> str | None:
    """The database URL for DB-backed tests, or None when none is configured.

    Unset returns None, and the callers turn that into a skip — running without
    a database is a supported state. Set-but-unsafe raises, because a skip would
    hide the misconfiguration rather than correct it.
    """
    value = (raw if raw is not None else os.environ.get("CONTROL_PLANE_TEST_DATABASE_URL") or "").strip()
    if not value:
        return None
    if is_test_database_url(value):
        return value
    raise UnsafeTestDatabaseError(_GUIDANCE.format(name=database_name_of(value) or "<none>"))

File: control-plane/api/test/support/test-database.ts
/**
 * The one place a test is allowed to learn which database it may destroy.
 *
 * Every DB-backed suite in this package opens with:
 *
 *     DROP SCHEMA IF EXISTS control_plane CASCADE
 *
 * which is correct for a scratch database and catastrophic for any other. On
 * 2026-09-06 `CONTROL_PLANE_TEST_DATABASE_URL` was pointed at
 * `control_plane_database_url_host` — the dev stack's OWN database, reached
 * through the host port rather than the compose network — and the suite wiped
 * the running control plane. The tests passed. Nothing warned. The stack was
 * only found broken afterwards, by its own readiness probe failing with 42P01.
 *
 * The convention that would have prevented it was already written down
 * (`docs/sprint5-trip-bot-router-design.md`: "tests drop and recreate the
 * schema, so point CONTROL_PLANE_TEST_DATABASE_URL at a scratch database").
 * A convention that only exists in prose is one every reader has to remember;
 * this module makes it a precondition instead, checked at the source rather
 * than at 25 call sites that would each have to remember to check.
 *
 * The rule is deliberately crude — the database NAME must say it is for tests
 * — because the failure it guards is crude. A subtler rule (is this port the
 * dev stack's? is this host local?) would have more ways to be wrong, and the
 * cost of being wrong here is the whole database.
 */

/** Names a database may have before this package will drop schemas in it. */
const TEST_NAME_PATTERN = /test/i;

export class UnsafeTestDatabaseError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "UnsafeTestDatabaseError";
  }
}

/**
 * The database name in a Postgres connection string, or null if there is none.
 *
 * Kept separate so the naming rule can be tested without constructing URLs
 * that look like credentials.
 */
export function databaseNameOf(connectionString: string): string | null {
  let parsed: URL;
  try {
    parsed = new URL(connectionString);
  } catch {
    return null;
  }
  const name = parsed.pathname.replace(/^\//, "");
  return name === "" ? null : decodeURIComponent(name);
}

/** Whether a connection string names a database this package may destroy. */
export function isTestDatabaseUrl(connectionString: string): boolean {
  const name = databaseNameOf(connectionString);
  return name !== null && TEST_NAME_PATTERN.test(name);
}

/**
 * The database URL for destructive suites, or `undefined` when none is set.
 *
 * Unset stays a SKIP — running the unit subset with no database is a normal,
 * supported state. Set-but-unsafe is a THROW, loudly and before any test runs:
 * a skip would hide the misconfiguration, and the next person would point it
 * at their stack too.
 */
export function testDatabaseUrl(
  raw: string | undefined = process.env.CONTROL_PLANE_TEST_DATABASE_URL,
): string | undefined {
  const value = (raw ?? "").trim();
  if (!value) return undefined;
  if (isTestDatabaseUrl(value)) return value;

  const name = databaseNameOf(value);
  throw new UnsafeTestDatabaseError(
    [
      `CONTROL_PLANE_TEST_DATABASE_URL names ${name ? `the database "${name}"` : "no database"},`,
      `which does not look like a scratch database.`,
      ``,
      `These suites begin by dropping the control_plane schema, so pointing them`,
      `at a real database destroys it. The dev stack's own database is exactly`,
      `the mistake this catches.`,
      ``,
      `Use the scratch database instead:`,
      `  CONTROL_PLANE_TEST_DATABASE_URL="postgres://postgres:test@127.0.0.1:5434/cptest"`,
      ``,
      `Or leave it unset to run the unit subset only.`,
    ].join("\n"),
  );
}
